languages/python/sqlalchemy-oso/tests/test_roles2.py

```diff
@@ -0,0 +1,146 @@
+
+from sqlalchemy_oso import register_models
+from sqlalchemy_oso.roles2 import OsoRoles
+
+from oso import Oso, Variable
+
+Base = declarative_base(name="RoleBase")
+
+
+class User(Base):
+    __tablename__ = "users"
+
+    id = Column(Integer, primary_key=True)
+    name = Column(String())
+
+
+class Organization(Base):
+    __tablename__ = "organizations"
+
+    id = Column(String(), primary_key=True)
+
+
+class Repository(Base):
+    __tablename__ = "repositories"
+
+    id = Column(String(), primary_key=True)
+    org_id = Column(String(), ForeignKey("organizations.id"))
+    org = relationship("Organization")
+
+
+class Issue(Base):
+    __tablename__ = "issues"
+
+    id = Column(String(), primary_key=True)
+    repo_id = Column(String(), ForeignKey("repositories.id"))
+    repo = relationship("Repository")
+
+
+def test_roles():
+    oso = Oso()
+    register_models(oso, Base)
+
+    roles = OsoRoles(oso, Base, User)
+    roles.enable()
+
+    policy = """
+    resource(_type: Organization, "org", actions, roles) if
+        actions = [
+            "invite",
+            "create_repo"
+        ] and
+        roles = {
+            org_member: {
+                perms: ["create_repo"],
+                implies: ["repo_read"]
+            },
+            org_owner: {
+                perms: ["invite"],
+                implies: ["org_member", "repo_write"]
+            }
+        };
+
+    resource(_type: Repository, "repo", actions, roles) if
+        actions = [
+            "push",
+            "pull"
+        ] and
+        roles = {
+            repo_write: {
+                perms: ["push", "issue:edit"],
+                implies: ["repo_read"]
+            },
+            repo_read: {
+                perms: ["pull"]
+            }
+        };
+        
+    resource(_type: Issue, "issue", actions, _) if
+        actions = [
+            "edit"
+        ];
+
+    parent(repository: Repository, parent_org: Organization) if
+        repository.org = parent_org;
+
+    parent(issue: Issue, parent_repo: Repository) if
+        issue.repo = parent_repo;
+
+    allow(actor, action, resource) if
+        Roles.role_allows(actor, action, resource);
+    """
+    oso.load_str(policy)
+
+    # engine = create_engine("sqlite:///roles.db")
+    engine = create_engine("sqlite://")
+    # @NOTE: Right now this has to happen after enabling oso roles to get the
+    #        tables.
+    Base.metadata.create_all(engine)
+
+    Session = sessionmaker(bind=engine)
+    session = Session()
+
+    osohq = Organization(id="osohq")
+    oso_repo = Repository(id="oso", org=osohq)
+    bug = Issue(id="bug", repo=oso_repo)
+
+    leina = User(name="leina")
+    steve = User(name="steve")
+
+    objs = [leina, steve, osohq, oso_repo, bug]
+    for obj in objs:
+        session.add(obj)
+    session.commit()
+
+    # @TODO: How the heck should this work?
+    # not like this...
+    roles.session = session
```

Comment:
This session is used within the policy when `is_allowed` is called?

Reply:
If we were just writing this entire library for SQLAlchemy I'd say this should just be passed into `is_allowed` as you've done for `assign_role`. Obviously that would not keep the API consistent `Oso` when you're not using SQLAlchemy.